File: leetcode/1535/solution.go
/**
 * https://leetcode.com/problems/find-the-winner-of-an-array-game/
 * 1535. Find the Winner of an Array Game
 */

package main

func getWinner(arr []int, k int) int {
	if len(arr) <= k {
		max := arr[0]
		for _, n := range arr {
			if max < n {
				max = n
			}
		}
		return max
	}

	count := 0
	i := 1
	winner := 0
	for count < k {
		if arr[winner] > arr[i] {
			count++
		} else if arr[winner] < arr[i] {
			count = 1
			winner = i
		}
		i = (i + 1) % len(arr)
	}

	return arr[winner]
}

File: leetcode/1535/solution.rs
/**
 * https://leetcode.com/problems/find-the-winner-of-an-array-game/
 * 1535. Find the Winner of an Array Game
 */

impl Solution {
    pub fn get_winner(arr: Vec<i32>, k: i32) -> i32 {
        // Shortcut
        if arr.len() < k as usize {
            return arr.into_iter().max().unwrap();
        }

        let (mut count, mut winner, mut i) = (0, 0, 0);
        while count < k {
            if arr[winner] > arr[i] {
                count += 1;
            } else if arr[winner] < arr[i] {
                count = 1;
                winner = i;
            }
            i = (i + 1) % arr.len();
        }

        arr[winner]
    }
}
